File: VB/HeadFirstPatterns/DoFactory.HeadFirst/DoFactory.HeadFirst.Iterator.DinerMergerI/MenuTestDrive.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections

Namespace DoFactory.HeadFirst.Iterator.DinerMergerI
	Friend Class MenuTestDrive
        Shared Sub Main(ByVal args() As String)

            Dim pancakeHouseMenu As New PancakeHouseMenu()
            Dim dinerMenu As New DinerMenu()

            Dim waitress As New Waitress(pancakeHouseMenu, dinerMenu)
            waitress.PrintMenu()
            waitress.PrintVegetarianMenu()

            Console.WriteLine(Constants.vbLf & "Customer asks, is the Hotdog vegetarian?")
            Console.Write("Waitress says: ")
            If waitress.IsItemVegetarian("Hotdog") Then
                Console.WriteLine("Yes")
            Else
                Console.WriteLine("No")
            End If

            Console.WriteLine(Constants.vbLf & "Customer asks, are the Waffles vegetarian?")
            Console.Write("Waitress says: ")
            If waitress.IsItemVegetarian("Waffles") Then
                Console.WriteLine("Yes")
            Else
                Console.WriteLine("No")
            End If

            ' Wait for user
            Console.ReadKey()
        End Sub
	End Class

	#Region "Waitress"

	Public Class Waitress
		Private _pancakeHouseMenu As Menu
		Private _dinerMenu As Menu

		Public Sub New(ByVal pancakeHouseMenu As Menu, ByVal dinerMenu As Menu)
			Me._pancakeHouseMenu = pancakeHouseMenu
			Me._dinerMenu = dinerMenu
		End Sub

		Public Sub PrintMenu()
			Dim pancakeIterator As IEnumerator = _pancakeHouseMenu.CreateIterator()
			Dim dinerIterator As IEnumerator = _dinerMenu.CreateIterator()

			Console.WriteLine("MENU" & Constants.vbLf & "----" & Constants.vbLf & "BREAKFAST")
			PrintMenu(pancakeIterator)

			Console.WriteLine(Constants.vbLf & "LUNCH")
			PrintMenu(dinerIterator)
		End Sub

		Private Sub PrintMenu(ByVal iterator As IEnumerator)
			Do While iterator.MoveNext()
				Dim menuItem As MenuItem = CType(iterator.Current, MenuItem)
				Console.Write(menuItem.Name & ", ")
				Console.Write(menuItem.Price & " -- ")
				Console.WriteLine(menuItem.Description)
			Loop
		End Sub

		Public Sub PrintVegetarianMenu()
			Console.WriteLine(Constants.vbLf & "VEGETARIAN MENU" & Constants.vbLf & "----" & Constants.vbLf & "BREAKFAST")
			PrintVegetarianMenu(_pancakeHouseMenu.CreateIterator())
			Console.WriteLine(Constants.vbLf & "LUNCH")
			PrintVegetarianMenu(_dinerMenu.CreateIterator())
		End Sub

		Public Function IsItemVegetarian(ByVal name As String) As Boolean
			Dim pancakeIterator As IEnumerator = _pancakeHouseMenu.CreateIterator()
			If IsVegetarian(name, pancakeIterator) Then
				Return True
			End If

			Dim dinerIterator As IEnumerator = _dinerMenu.CreateIterator()
			If IsVegetarian(name, dinerIterator) Then
				Return True
			End If
			Return False
		End Function

		Private Sub PrintVegetarianMenu(ByVal iterator As IEnumerator)
			Do While iterator.MoveNext()
				Dim menuItem As MenuItem = CType(iterator.Current, MenuItem)
				If menuItem.Vegetarian Then
					Console.Write(menuItem.Name & ", ")
					Console.Write(menuItem.Price & " -- ")
					Console.WriteLine(menuItem.Description)
				End If
			Loop
		End Sub

		Private Function IsVegetarian(ByVal name As String, ByVal iterator As IEnumerator) As Boolean
			Do While iterator.MoveNext()
				Dim menuItem As MenuItem = CType(iterator.Current, MenuItem)
				If menuItem.Name.Equals(name) Then
					If menuItem.Vegetarian Then
						Return True
					End If
				End If
			Loop
			Return False
		End Function
	End Class

	#End Region

	#Region "Iterators (Enumerators)"

	Public Class AlternatingDinerMenuIterator
		Implements IEnumerator
		Private _items() As MenuItem
		Private _position As Integer

		Public Sub New(ByVal items() As MenuItem)
			Me._items = items
			_position = Integer.Parse(DateTime.Now.DayOfWeek.ToString()) Mod 2
		End Sub

		Public ReadOnly Property Current() As Object Implements IEnumerator.Current
			Get
				Dim menuItem As MenuItem = _items(_position)
				_position = _position + 2
				Return menuItem
			End Get
		End Property

		Public Function MoveNext() As Boolean Implements IEnumerator.MoveNext
			If _position >= _items.Length OrElse _items(_position) Is Nothing Then
				Return False
			Else
				Return True
			End If
		End Function

		Public Sub Reset() Implements IEnumerator.Reset
			Console.WriteLine("Alternating Diner Menu Iterator does not support reset()")
		End Sub
	End Class

	Public Class DinerMenuIterator
		Implements IEnumerator
		Private _list() As MenuItem
		Private _position As Integer = 0

		Public Sub New(ByVal list() As MenuItem)
			Me._list = list
		End Sub

		Public ReadOnly Property Current() As Object Implements IEnumerator.Current
			Get

				Dim menuItem As MenuItem = _list(_position)
				_position = _position + 1
				Return menuItem
			End Get
		End Property

		Public Function MoveNext() As Boolean Implements IEnumerator.MoveNext
			If _position >= _list.Length OrElse _list(_position) Is Nothing Then
				Return False
			Else
				Return True
			End If
		End Function

		Public Sub Reset() Implements IEnumerator.Reset
			If _position <= 0 Then
				Throw New ApplicationException ("You can't remove an item until you've done at least one Next()")
			End If
			If _list(_position - 1) IsNot Nothing Then
				For i As Integer = _position - 1 To (_list.Length - 1) - 1
					_list(i) = _list(i + 1)
				Next i
				_list(_list.Length - 1) = Nothing
			End If
		End Sub
	End Class

	#End Region

	#Region "Menu and MenuItems"

	Public Interface Menu
		Function CreateIterator() As IEnumerator
	End Interface

	Public Class MenuItem
		Private privateName As String
		Public Property Name() As String
			Get
				Return privateName
			End Get
			Set(ByVal value As String)
				privateName = value
			End Set
		End Property
		Private privateDescription As String
		Public Property Description() As String
			Get
				Return privateDescription
			End Get
			Set(ByVal value As String)
				privateDescription = value
			End Set
		End Property
		Private privateVegetarian As Boolean
		Public Property Vegetarian() As Boolean
			Get
				Return privateVegetarian
			End Get
			Set(ByVal value As Boolean)
				privateVegetarian = value
			End Set
		End Property
		Private privatePrice As Double
		Public Property Price() As Double
			Get
				Return privatePrice
			End Get
			Set(ByVal value As Double)
				privatePrice = value
			End Set
		End Property

		' Constructor
		Public Sub New(ByVal name As String, ByVal description As String, ByVal vegetarian As Boolean, ByVal price As Double)
			Me.Name = name
			Me.Description = description
			Me.Vegetarian = vegetarian
			Me.Price = price
		End Sub
	End Class

	Public Class DinerMenu
		Implements Menu
		Private Shared ReadOnly MAX_ITEMS As Integer = 6
		Private _numberOfItems As Integer = 0
		Private _menuItems(MAX_ITEMS - 1) As MenuItem

		Public Sub New()
			AddItem("Vegetarian BLT", "(Fakin') Bacon with lettuce & tomato on whole wheat", True, 2.99)
			AddItem("BLT", "Bacon with lettuce & tomato on whole wheat", False, 2.99)
			AddItem("Soup of the day", "Soup of the day, with a side of potato salad", False, 3.29)
			AddItem("Hotdog", "A hot dog, with saurkraut, relish, onions, topped with cheese", False, 3.05)
			AddItem("Steamed Veggies and Brown Rice", "Steamed vegetables over brown rice", True, 3.99)
			AddItem("Pasta", "Spaghetti with Marinara Sauce, and a slice of sourdough bread", True, 3.89)
		End Sub

		Public Sub AddItem(ByVal name As String, ByVal description As String, ByVal vegetarian As Boolean, ByVal price As Double)
			Dim menuItem As New MenuItem(name, description, vegetarian, price)
			If _numberOfItems >= MAX_ITEMS Then
				Console.WriteLine("Sorry, menu is full!  Can't add item to menu")
			Else
				_menuItems(_numberOfItems) = menuItem
				_numberOfItems = _numberOfItems + 1
			End If
		End Sub

		Public Function getMenuItems() As MenuItem()
			Return _menuItems
		End Function

		Public Function CreateIterator() As IEnumerator Implements Menu.CreateIterator
			Return New DinerMenuIterator(_menuItems)
		End Function

		' other menu methods here
	End Class


	Public Class PancakeHouseMenu
		Implements Menu
		Private _menuItems As ArrayList

		Public Sub New()
			_menuItems = New ArrayList()

			AddItem("K&B's Pancake Breakfast", "Pancakes with scrambled eggs, and toast", True, 2.99)

			AddItem("Regular Pancake Breakfast", "Pancakes with fried eggs, sausage", False, 2.99)

			AddItem("Blueberry Pancakes", "Pancakes made with fresh blueberries, and blueberry syrup", True, 3.49)

			AddItem("Waffles", "Waffles, with your choice of blueberries or strawberries", True, 3.59)
		End Sub

		Public Sub AddItem(ByVal name As String, ByVal description As String, ByVal vegetarian As Boolean, ByVal price As Double)
			Dim menuItem As New MenuItem(name, description, vegetarian, price)
			_menuItems.Add(menuItem)
		End Sub

		Public Function getMenuItems() As ArrayList
			Return _menuItems
		End Function

		Public Function CreateIterator() As IEnumerator Implements Menu.CreateIterator
			Return _menuItems.GetEnumerator()
		End Function

		' other menu methods here
	End Class

	#End Region
End Namespace

File: CSharp/HeadFirstPatterns/DoFactory.HeadFirst/DoFactory.HeadFirst.Iterator.DinerMergerI/MenuTestDrive.cs
using System;
using System.Collections;

namespace DoFactory.HeadFirst.Iterator.DinerMergerI
{
    class MenuTestDrive
    {
        static void Main(string[] args)
        {
            PancakeHouseMenu pancakeHouseMenu = new PancakeHouseMenu();
            DinerMenu dinerMenu = new DinerMenu();

            Waitress waitress = new Waitress(pancakeHouseMenu, dinerMenu);
            waitress.PrintMenu();
            waitress.PrintVegetarianMenu();

            Console.WriteLine("\nCustomer asks, is the Hotdog vegetarian?");
            Console.Write("Waitress says: ");
            if (waitress.IsItemVegetarian("Hotdog"))
            {
                Console.WriteLine("Yes");
            }
            else
            {
                Console.WriteLine("No");
            }

            Console.WriteLine("\nCustomer asks, are the Waffles vegetarian?");
            Console.Write("Waitress says: ");
            if (waitress.IsItemVegetarian("Waffles"))
            {
                Console.WriteLine("Yes");
            }
            else
            {
                Console.WriteLine("No");
            }

            // Wait for user
            Console.ReadKey();
        }
    }

    #region Waitress

    public class Waitress
    {
        private Menu _pancakeHouseMenu;
        private Menu _dinerMenu;

        public Waitress(Menu pancakeHouseMenu, Menu dinerMenu)
        {
            this._pancakeHouseMenu = pancakeHouseMenu;
            this._dinerMenu = dinerMenu;
        }

        public void PrintMenu()
        {
            IEnumerator pancakeIterator = _pancakeHouseMenu.CreateIterator();
            IEnumerator dinerIterator = _dinerMenu.CreateIterator();

            Console.WriteLine("MENU\n----\nBREAKFAST");
            PrintMenu(pancakeIterator);

            Console.WriteLine("\nLUNCH");
            PrintMenu(dinerIterator);
        }

        private void PrintMenu(IEnumerator iterator)
        {
            while (iterator.MoveNext())
            {
                MenuItem menuItem = (MenuItem)iterator.Current;
                Console.Write(menuItem.Name + ", ");
                Console.Write(menuItem.Price + " -- ");
                Console.WriteLine(menuItem.Description);
            }
        }

        public void PrintVegetarianMenu()
        {
            Console.WriteLine("\nVEGETARIAN MENU\n----\nBREAKFAST");
            PrintVegetarianMenu(_pancakeHouseMenu.CreateIterator());
            Console.WriteLine("\nLUNCH");
            PrintVegetarianMenu(_dinerMenu.CreateIterator());
        }

        public bool IsItemVegetarian(String name)
        {
            IEnumerator pancakeIterator = _pancakeHouseMenu.CreateIterator();
            if (IsVegetarian(name, pancakeIterator))
            {
                return true;
            }

            IEnumerator dinerIterator = _dinerMenu.CreateIterator();
            if (IsVegetarian(name, dinerIterator))
            {
                return true;
            }
            return false;
        }

        private void PrintVegetarianMenu(IEnumerator iterator)
        {
            while (iterator.MoveNext())
            {
                MenuItem menuItem = (MenuItem)iterator.Current;
                if (menuItem.Vegetarian)
                {
                    Console.Write(menuItem.Name + ", ");
                    Console.Write(menuItem.Price + " -- ");
                    Console.WriteLine(menuItem.Description);
                }
            }
        }

        private bool IsVegetarian(string name, IEnumerator iterator)
        {
            while (iterator.MoveNext())
            {
                MenuItem menuItem = (MenuItem)iterator.Current;
                if (menuItem.Name.Equals(name))
                {
                    if (menuItem.Vegetarian)
                    {
                        return true;
                    }
                }
            }
            return false;
        }
    }

    #endregion

    #region Iterators (Enumerators)

    public class AlternatingDinerMenuIterator : IEnumerator
    {
        private MenuItem[] _items;
        private int _position;

        public AlternatingDinerMenuIterator(MenuItem[] items)
        {
            this._items = items;
            _position = int.Parse(DateTime.Now.DayOfWeek.ToString()) % 2;
        }

        public object Current
        {
            get
            {
                MenuItem menuItem = _items[_position];
                _position = _position + 2;
                return menuItem;
            }
        }

        public bool MoveNext()
        {
            if (_position >= _items.Length || _items[_position] == null)
            {
                return false;
            }
            else
            {
                return true;
            }
        }

        public void Reset()
        {
            Console.WriteLine("Alternating Diner Menu Iterator does not support reset()");
        }
    }

    public class DinerMenuIterator : IEnumerator
    {
        private MenuItem[] _list;
        private int _position = 0;

        public DinerMenuIterator(MenuItem[] list)
        {
            this._list = list;
        }

        public object Current
        {
            get
            {

                MenuItem menuItem = _list[_position];
                _position = _position + 1;
                return menuItem;
            }
        }

        public bool MoveNext()
        {
            if (_position >= _list.Length || _list[_position] == null)
            {
                return false;
            }
            else
            {
                return true;
            }
        }

        public void Reset()
        {
            if (_position <= 0)
            {
                throw new ApplicationException
                    ("You can't remove an item until you've done at least one Next()");
            }
            if (_list[_position - 1] != null)
            {
                for (int i = _position - 1; i < (_list.Length - 1); i++)
                {
                    _list[i] = _list[i + 1];
                }
                _list[_list.Length - 1] = null;
            }
        }
    }

    #endregion

    #region Menu and MenuItems

    public interface Menu
    {
        IEnumerator CreateIterator();
    }

    public class MenuItem
    {
        public string Name { get; set; }
        public string Description { get; set; }
        public bool Vegetarian { get; set; }
        public double Price { get; set; }

        // Constructor
        public MenuItem(string name, string description, bool vegetarian, double price)
        {
            this.Name = name;
            this.Description = description;
            this.Vegetarian = vegetarian;
            this.Price = price;
        }
    }

    public class DinerMenu : Menu
    {
        private static readonly int MAX_ITEMS = 6;
        private int _numberOfItems = 0;
        private MenuItem[] _menuItems = new MenuItem[MAX_ITEMS];

        public DinerMenu()
        {
            AddItem("Vegetarian BLT",
                "(Fakin') Bacon with lettuce & tomato on whole wheat", true, 2.99);
            AddItem("BLT",
                "Bacon with lettuce & tomato on whole wheat", false, 2.99);
            AddItem("Soup of the day",
                "Soup of the day, with a side of potato salad", false, 3.29);
            AddItem("Hotdog",
                "A hot dog, with saurkraut, relish, onions, topped with cheese",
                false, 3.05);
            AddItem("Steamed Veggies and Brown Rice",
                "Steamed vegetables over brown rice", true, 3.99);
            AddItem("Pasta",
                "Spaghetti with Marinara Sauce, and a slice of sourdough bread",
                true, 3.89);
        }

        public void AddItem(string name, string description, bool vegetarian, double price)
        {
            MenuItem menuItem = new MenuItem(name, description, vegetarian, price);
            if (_numberOfItems >= MAX_ITEMS)
            {
                Console.WriteLine("Sorry, menu is full!  Can't add item to menu");
            }
            else
            {
                _menuItems[_numberOfItems] = menuItem;
                _numberOfItems = _numberOfItems + 1;
            }
        }

        public MenuItem[] getMenuItems()
        {
            return _menuItems;
        }

        public IEnumerator CreateIterator()
        {
            return new DinerMenuIterator(_menuItems);
        }

        // other menu methods here
    }


    public class PancakeHouseMenu : Menu
    {
        private ArrayList _menuItems;

        public PancakeHouseMenu()
        {
            _menuItems = new ArrayList();

            AddItem("K&B's Pancake Breakfast",
                "Pancakes with scrambled eggs, and toast",
                true,
                2.99);

            AddItem("Regular Pancake Breakfast",
                "Pancakes with fried eggs, sausage",
                false,
                2.99);

            AddItem("Blueberry Pancakes",
                "Pancakes made with fresh blueberries, and blueberry syrup",
                true,
                3.49);

            AddItem("Waffles",
                "Waffles, with your choice of blueberries or strawberries",
                true,
                3.59);
        }

        public void AddItem(String name, string description, bool vegetarian, double price)
        {
            MenuItem menuItem = new MenuItem(name, description, vegetarian, price);
            _menuItems.Add(menuItem);
        }

        public ArrayList getMenuItems()
        {
            return _menuItems;
        }

        public IEnumerator CreateIterator()
        {
            return _menuItems.GetEnumerator();
        }

        // other menu methods here
    }

    #endregion
}
